truffle/src/main/java/org/jruby/truffle/runtime/rope/RopeOperations.java

```diff
@@ -144,4 +131,216 @@ public static int strLength(Encoding enc, byte[] bytes, int p, int end) {
         return StringSupport.strLength(enc, bytes, p, end);
     }
 
+    public static LeafRope flatten(Rope rope) {
+        if (rope instanceof LeafRope) {
+            return (LeafRope) rope;
+        }
+
+        return create(flattenBytes(rope), rope.getEncoding(), rope.getCodeRange());
+    }
+
+    @TruffleBoundary
+    /**
+     * Performs an iterative depth first search of the Rope tree to calculate its byte[] without needing to populate
+     * the byte[] for each level beneath. Every LeafRope has its byte[] populated by definition. The goal is to determine
+     * which descendant LeafRopes contribute bytes to the top-most Rope's logical byte[] and how many bytes they should
+     * contribute. Then each such LeafRope copies the appropriate range of bytes to a shared byte[].
+     *
+     * Rope trees can be very deep. An iterative algorithm is preferable to recursion because it removes the overhead
+     * of stack frame management. Additionally, a recursive algorithm will eventually overflow the stack if the Rope
+     * tree is too deep.
+     */
+    public static byte[] flattenBytes(Rope rope) {
```

Comment:
The recursive solution was a lot slower and can never PE. I think in reality this is too complex for PE anyway, but the option is available. But, currently ropes aren't bounded in depth. I'm still trying to find a good heuristic for limiting depth, but the ones I've tried thus far have mostly resulted in large performance regressions. It may be the case that it was a combination of the slower recursive function and the increased call frequency. But I was generally finding things were a lot faster and cheaper with memory if I let the Rope get arbitrarily long and took a single pass across the tree once the bytes were actually needed.